### PR TITLE
[WIP][SPARK-6986][CORE]Make SerializationStream/DeserializationStream understand key/value semantic

### PR DESCRIPTION
**Pull request by yhuai**
https://issues.apache.org/jira/browse/SPARK-6986

### CODE REVIEW

File: core/src/main/scala/org/apache/spark/util/collection/ExternalSorter.scala [hunk 1]
```diff
@@ -300,8 +300,8 @@ private[spark] class ExternalSorter[K, V, C](
         val partitionId = elem._1._1
         val key = elem._1._2
         val value = elem._2
-        writer.write(key)
-        writer.write(value)
+        writer.writeKey(key)
+        writer.writeValue(value)
```
Comment (yhuai) [inline core/src/main/scala/org/apache/spark/util/collection/ExternalSorter.scala]: @rxin An alternative way to make the specialized SQL serializer work in most of the cases is to only change 
```
writer.write(key)
writer.write(value)
```
to 
```
writer.write((key, value))
```

**Comment by sryza** [comment]
My patch for SPARK-4550 ( #4450 ) makes a similar change.

**Comment by pwendell** [comment]
Seems good to me - @rxin any comments?

**Comment by sryza** [comment]
As it bumps numRecordsWritten both on key and on value, it looks like this change would make us double count the number of records written.

**Comment by yhuai** [comment]
@sryza Without this change, we still have this issue (when we call `write(key)` and then `write(value)` separately), right? Do you have any suggestion on how to address it?

**Comment by yhuai** [comment]
@sryza's commit https://github.com/apache/spark/commit/0a2b15ce43cf6096e1a7ae060b7c8a4010ce3b92 already exposes the interfaces. I am closing it

**Pull request closed without merging**